### ISSUE TITLE
Inconsistent output: Closure Compiler Service vs google-closure-compiler

### ISSUE THREAD

**Issue by christiantjl** [google/closure-compiler#3608]
Working on #3035, using ADVANCED compilation mode.

Input:
```
(function (window) {
        
    /* inject CC output for object-assign.js */
    var k="function"==typeof Object.defineProperties?Object.defineProperty:function(a,d,c){a!=Array.prototype&&a!=Object.prototype&&(a[d]=c.value)},l="undefined"!=typeof window&&window===this?this:"undefined"!=typeof global&&null!=global?global:this;function m(a,d){if(d){var c=l;a=a.split(".");for(var b=0;b<a.length-1;b++){var e=a[b];e in c||(c[e]={});c=c[e]}a=a[a.length-1];b=c[a];d=d(b);d!=b&&null!=d&&k(c,a,{configurable:!0,writable:!0,value:d})}}
    m("Object.getOwnPropertySymbols",function(a){return a?a:function(){return[]}});var n="function"==typeof Object.assign?Object.assign:function(a,d){for(var c=1;c<arguments.length;c++){var b=arguments[c];if(b)for(var e in b)Object.prototype.hasOwnProperty.call(b,e)&&(a[e]=b[e])}return a};m("Object.assign",function(a){return a||n});
    var p=Object.getOwnPropertySymbols,q=Object.prototype.hasOwnProperty,r=Object.prototype.propertyIsEnumerable,t=function(){try{if(!Object.assign)return!1;var a=new String("abc");a[5]="de";if("5"===Object.getOwnPropertyNames(a)[0])return!1;var d={};for(a=0;10>a;a++)d["_"+String.fromCharCode(a)]=a;if("0123456789"!==Object.getOwnPropertyNames(d).map(function(b){return d[b]}).join(""))return!1;var c={};"abcdefghijklmnopqrst".split("").forEach(function(b){c[b]=b});return"abcdefghijklmnopqrst"!==Object.keys(Object.assign({},
    c)).join("")?!1:!0}catch(b){return!1}}()?Object.assign:function(a,d){if(null===a||void 0===a)throw new TypeError("Object.assign cannot be called with null or undefined");var c=Object(a);for(var b,e=1;e<arguments.length;e++){var f=Object(arguments[e]);for(var h in f)q.call(f,h)&&(c[h]=f[h]);if(p){b=p(f);for(var g=0;g<b.length;g++)r.call(f,b[g])&&(c[b[g]]=f[b[g]])}}return c};window.__NODE_MODULE__OBJECT_ASSIGN=t;

    /* end CC inject */
    var __module = (typeof module === "undefined" ? {} : module); 
    __module.exports = window.__NODE_MODULE__OBJECT_ASSIGN;
    return window.__NODE_MODULE__OBJECT_ASSIGN;

})(this)
```

Output - **CORRECT** (https://closure-compiler.appspot.com/):
```
(function(h){function l(a,d){if(d){var c=n;a=a.split(".");for(var b=0;b<a.length-1;b++){var e=a[b];e in c||(c[e]={});c=c[e]}a=a[a.length-1];b=c[a];d=d(b);d!=b&&null!=d&&p(c,a,{configurable:!0,writable:!0,value:d})}}var p="function"==typeof Object.defineProperties?Object.defineProperty:function(a,d,c){a!=Array.prototype&&a!=Object.prototype&&(a[d]=c.value)},n="undefined"!=typeof h&&h===this?this:"undefined"!=typeof global&&null!=global?global:this;l("Object.getOwnPropertySymbols",function(a){return a?
a:function(){return[]}});var q="function"==typeof Object.assign?Object.assign:function(a,d){for(var c=1;c<arguments.length;c++){var b=arguments[c];if(b)for(var e in b)Object.prototype.hasOwnProperty.call(b,e)&&(a[e]=b[e])}return a};l("Object.assign",function(a){return a||q});var m=Object.getOwnPropertySymbols,r=Object.prototype.hasOwnProperty,t=Object.prototype.propertyIsEnumerable,u=function(){try{if(!Object.assign)return!1;var a=new String("abc");a[5]="de";if("5"===Object.getOwnPropertyNames(a)[0])return!1;
var d={};for(a=0;10>a;a++)d["_"+String.fromCharCode(a)]=a;if("0123456789"!==Object.getOwnPropertyNames(d).map(function(b){return d[b]}).join(""))return!1;var c={};"abcdefghijklmnopqrst".split("").forEach(function(b){c[b]=b});return"abcdefghijklmnopqrst"!==Object.keys(Object.assign({},c)).join("")?!1:!0}catch(b){return!1}}()?Object.assign:function(a,d){if(null===a||void 0===a)throw new TypeError("Object.assign cannot be called with null or undefined");for(var c=Object(a),b,e=1;e<arguments.length;e++){var f=
Object(arguments[e]),k;for(k in f)r.call(f,k)&&(c[k]=f[k]);if(m){b=m(f);for(var g=0;g<b.length;g++)t.call(f,b[g])&&(c[b[g]]=f[b[g]])}}return c};h.a=u;return("undefined"===typeof module?{}:module).exports=h.a})(this);
```
Output - **USELESS** (`google-closure-compiler --compilation_level ADVANCED_OPTIMIZATIONS`):
```
var q="function"==typeof Object.defineProperties?Object.defineProperty:function(b,h,e){b!=Array.prototype&&b!=Object.prototype&&(b[h]=e.value)},t="undefined"!=typeof window&&window===this?this:"undefined"!=typeof global&&null!=global?global:this;function u(b,h){if(h){var e=t;b=b.split(".");for(var g=0;g<b.length-1;g++){var l=b[g];l in e||(e[l]={});e=e[l]}b=b[b.length-1];g=e[b];h=h(g);h!=g&&null!=h&&q(e,b,{configurable:!0,writable:!0,value:h})}}
u("Object.getOwnPropertySymbols",function(b){return b?b:function(){return[]}});var v="function"==typeof Object.assign?Object.assign:function(b,h){for(var e=1;e<arguments.length;e++){var g=arguments[e];if(g)for(var l in g)Object.prototype.hasOwnProperty.call(g,l)&&(b[l]=g[l])}return b};u("Object.assign",function(b){return b||v});
(function(b){function h(a,f){if(f){var d=g;a=a.split(".");for(var c=0;c<a.length-1;c++){var k=a[c];k in d||(d[k]={});d=d[k]}a=a[a.length-1];c=d[a];f=f(c);f!=c&&null!=f&&e(d,a,{configurable:!0,writable:!0,value:f})}}var e="function"==typeof Object.defineProperties?Object.defineProperty:function(a,f,d){a!=Array.prototype&&a!=Object.prototype&&(a[f]=d.value)},g="undefined"!=typeof b&&b===this?this:"undefined"!=typeof global&&null!=global?global:this;h("Object.getOwnPropertySymbols",function(a){return a?
a:function(){return[]}});var l="function"==typeof Object.assign?Object.assign:function(a,f){for(var d=1;d<arguments.length;d++){var c=arguments[d];if(c)for(var k in c)Object.prototype.hasOwnProperty.call(c,k)&&(a[k]=c[k])}return a};h("Object.assign",function(a){return a||l});var r=Object.getOwnPropertySymbols,w=Object.prototype.hasOwnProperty,x=Object.prototype.propertyIsEnumerable,y=function(){try{if(!Object.assign)return!1;var a=new String("abc");a[5]="de";if("5"===Object.getOwnPropertyNames(a)[0])return!1;
var f={};for(a=0;10>a;a++)f["_"+String.fromCharCode(a)]=a;if("0123456789"!==Object.getOwnPropertyNames(f).map(function(c){return f[c]}).join(""))return!1;var d={};"abcdefghijklmnopqrst".split("").forEach(function(c){d[c]=c});return"abcdefghijklmnopqrst"!==Object.keys(Object.assign({},d)).join("")?!1:!0}catch(c){return!1}}()?Object.assign:function(a,f){if(null===a||void 0===a)throw new TypeError("Object.assign cannot be called with null or undefined");for(var d=Object(a),c,k=1;k<arguments.length;k++){var m=
Object(arguments[k]),p;for(p in m)w.call(m,p)&&(d[p]=m[p]);if(r){c=r(m);for(var n=0;n<c.length;n++)x.call(m,c[n])&&(d[c[n]]=m[c[n]])}}return d};b.a=y;return("undefined"===typeof module?{}:module).b=b.a})(this);
```

Any help resolving this difference would be optimal. The REST version is actually what I need - the current version of the compiler unwraps my anonymous function (c'mon) and breaks `require` by deciding to rename the `exports` property.

`require` with CORRECT output:
```
$ node -e "console.log(require('./compiled/object-assign-include.js'))"
> [Function: assign]
```

`require` with USELESS output:
```
node -e "console.log(require('./compiled/object-assign-include.js'))"
> { a: [Function: assign] }
```

**Comment by brad4d** [comment]
After running through clang-format and diffing I can see that the only real difference between the two besides the choices of variable names are:

From the web server:
```javascript
(function(h) {
  // main content
  return ('undefined' === typeof module ? {} : module).exports = h.a;
})(this);
```

From the command line:

```javascript
// polyfills here for Object.assign & friends
(function(h) {
// main content
  return ('undefined' === typeof module ? {} : module).h = h.a;
})(this);
```

So, the IIFE is not actually unwrapped. It just appeared so because of the polyfills that got injected in front of it.

I'm not sure of the whole reason why the `exports` got renamed on the command line but not from the web server, but I'm sure it's related to the compiler being unsure what kind of input module it was looking at (goog.module / common JS module / ES module).

It looks like the intention here was to wrap code already sent through the compiler and send it through the compiler again. That is explicitly not supported.

Perhaps if you could give us an idea of what you were trying to accomplish we could be more helpful.

**Comment by ChadKillingsworth** [comment]
The web-service runs in 3rd party mode and assumes undeclared names are external. That's the difference.

To prevent renaming, you need an extern for `module.exports`:

```js
/** @externs */

const module = {};
/** @type {?} */
module.exports;
```

**Comment by christiantjl** [comment]
Many thanks, I will close this issue.

@brad4d Slightly off-topic, but I listened to a speech you gave on Closure Compiler and I learned a lot from it. I’m pretty out of my depth, but I’d be very willing to devote time and effort to help get GCC to a point where it supports class static fields, though I know you mentioned there were other things to do first (BigInt etc). If there’s any way I could help, or if there’s some big TODO list somewhere, my email is visible on my profile. =)
